src/components/validation/runner/mod.rs

```diff
@@ -483,6 +510,83 @@ fn spawn_component_topology(
     });
 }
 
+fn spawn_input_driver(
```

Comment:
This is a lot cleaner overall, nice.

One random nit: we'll want to be careful that the codec emits _all_ necessary bytes for the payload. Right now, as an example, for the HTTP resource, we're tracking the payload size because we only care about the body size... and the `http_server` source only tracks the body payload bytes, I guess. That's all fine and good, and matches how we're using the encoded size of the event here as the expected size.

However, you could imagine a scenario where maybe we track literally _every_ byte that comes in over the HTTP connection -- request headers and all -- and then this approach would break down.

I can tell you that we definitely do that sort of approach for some components, such as the `vector` source, where we wrap the connection and count the literal bytes that flowed over the TCP socket. With this approach, and having a naive encoder implementation that simply encoded the event to its binary Protocol Buffers form... we'd end up with a byte size mismatch.

It might end up being easier at some point to move the collecting of these metrics to the resources themselves so that they can account for that, but I think this is a good step for right now because I do believe this logic will cover the majority of components.

Reply:
That's good to know ... I considered putting this in the resources themselves, but as I wasn't thinking of this concern, it seemed better at the time to put it where I did to simplify it and not require each resource to do it.

Will keep this in mind though going forward.

Reply:
Hmm. I started working on the sink side validators and ran into this with the `http` sink `SentBytesTotal` . Per the component spec 

```
For HTTP-based protocols, the total number of bytes in the HTTP body, as represented by the Content-Length header.
```

I checked and indeed in the external resource it's got the correct byte count.

Unless I'm missing something, it seems like we're going to have to push this logic into the external resources now.

Reply:
Ah yeah, not too surprising. 😬